drivers/serial/uart_nrfx_uart.c

```diff
@@ -73,8 +73,10 @@ static struct {
 	size_t rx_secondary_buffer_length;
 	volatile size_t rx_counter;
 	volatile size_t rx_offset;
+#ifdef CONFIG_SYS_CLOCK_EXISTS
 	int32_t rx_timeout;
 	struct k_timer rx_timeout_timer;
+#endif
```

Comment:
This is not a proper way of solving the compilation issue. Without timers, the asynchronous API will not be handled correctly in the nRF UART drivers, so users should not be able to enable this API in such case.
Instead of adding all these `#ifdef`s, you should rather change: https://github.com/zephyrproject-rtos/zephyr/blob/75a2ef8fc02a417f0d26307105f80b136a666dc8/drivers/serial/Kconfig.nrfx#L11 to:
```
	select SERIAL_SUPPORT_ASYNC if SYS_CLOCK_EXISTS
```

Reply:
@anangl yes, api will have limited use, on tx side there will be no timeout when hwfc is enabled. On async there will be no events on partial data. Only when full buffer is received. Why not use interrupt driven api then if you have no sys clock? What would be the benefits of async api?

Reply:
My use case is the following one : uart is receiving packets of data and so I perfectly know the expected size I will receive. So in that case there is no need to have partial data. I only need to be notified when the full message is received. Using interrupt api will simply consume cpu power with dozens of intermediate interrupts (total size/hardware fifo size), instead of relying on hardware doing the job for the full packet (which is the whole point of easy dma I think).

So I think there is a usage for DMA-driven async uarte without the need of any timeout.
The behavior is the same when you specify SYS_FOREVER_MS as timeout value, so this looked an anticipated behavior anyway.

Reply:
@nordic-krch I agree that the commit looks a bit scary like this (it is better when you look at the complete file though). Indeed I tried my best to keep the existing style. Here is sample extract from existing code :

```
/* Device data structure */
struct uarte_nrfx_data {
	const struct device *dev;
	struct uart_config uart_config;
#ifdef UARTE_INTERRUPT_DRIVEN
	struct uarte_nrfx_int_driven *int_driven;
#endif
#ifdef CONFIG_UART_ASYNC_API
	struct uarte_async_cb *async;
#endif
	atomic_val_t poll_out_lock;
#ifdef CONFIG_PM_DEVICE
	uint32_t pm_state;
#endif
	uint8_t char_out;
	uint8_t rx_data;
	gppi_channel_t ppi_ch_endtx;
};
```

Pattern seemed very similar. And since we are targeting embedded device with limited RAM I guess it still makes sense though ?

Feedback about nrfx_glue is a very good point, as this specific part of code has a broader scope than just uart.
I will split it as soon as there is a consensus on the global usefulness of async without timeouts.

As this is my first contribution to zephyr I am not sure how to properly create a test case. I will try to figure out.

Reply:
@mspeder I see that I misunderstood the intention of this PR in the first place. Yes, the use case you provided is reasonable (perhaps it would be good to mention it in the commit message). I agree with @nordic-krch that that many `#ifdef`s is not nice looking at all, but I guess we'll need to live with it. I'd only suggest to use a definition like:
```
#define TIMEOUTS_AVAILABLE  IS_ENABLED(CONFIG_SYS_CLOCK_EXISTS)
```
to make the code more self-explanatory (`#ifdef CONFIG_SYS_CLOCK_EXISTS` is a bit cryptic).

And since the asynchronous API allows specifying timeouts, if it is to be supported even if timeouts are not available, `uart[e]_nrfx_tx` and `uart[e]_nrfx_rx_enable` should return `-ENOTSUP` (and perhaps provide some `LOG_ERR` message to make it clear what's wrong) when a timeout value other than `SYS_FOREVER_MS` is used.